oidc_middleware.go

```diff
@@ -0,0 +1,87 @@
+package main
+
+import (
+	"crypto/rsa"
+	"encoding/base64"
+	"encoding/json"
+	"github.com/gin-gonic/gin"
+	"math/big"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/golang-jwt/jwt"
+)
+// Largely taken from https://developer.okta.com/blog/2021/01/04/offline-jwt-validation-with-go
+
+func GetOidcPublicKeys(config OICDConfig) *map[string]*rsa.PublicKey {
+	rsakeys := make(map[string]*rsa.PublicKey)
+	var body map[string]interface{}
+	resp, err := http.Get(config.KeysURL)
+	if err != nil {
+		log.Error().WithError(err).Message("Could not fetch from KeysURL.")
+	}
+	err = json.NewDecoder(resp.Body).Decode(&body)
+	if err != nil {
+		log.Error().WithError(err).Message("Failed to decode login JWT Keys.")
+	}
+	log.Debug().Message("Updating keys for oidc.")
+	for _, bodykey := range body["keys"].([]interface{}) {
+		key := bodykey.(map[string]interface{})
+		kid := key["kid"].(string)
+		rsakey := new(rsa.PublicKey)
+		number, _ := base64.RawURLEncoding.DecodeString(key["n"].(string))
+		rsakey.N = new(big.Int).SetBytes(number)
+		rsakey.E = 65537
+		rsakeys[kid] = rsakey
+	}
+	return &rsakeys
+}
+
+func VerifyTokenMiddleware(config OICDConfig, rsakeys *map[string]*rsa.PublicKey) gin.HandlerFunc {
+	return func (ginContext *gin.Context) {
+		if *rsakeys == nil {
+			log.Warn().Message("RsaKeys for OIDC have not been set (http:500).")
+			ginContext.AbortWithStatus(http.StatusInternalServerError)
+		}
+		isValid := false
+		errorMessage := ""
+		tokenString := ginContext.Request.Header.Get("Authorization")
+		if strings.HasPrefix(tokenString, "Bearer ") {
+			tokenString = strings.TrimPrefix(tokenString, "Bearer ")
+			token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
+				return (*rsakeys)[token.Header["kid"].(string)], nil
```

Comment:
Neither the cast nor the map access is checked and are prone to panics.

Could you change these to do safe casting and safe map access?

```go
value := myMap["foo"]     // unsafe, panics if not found
value, ok := myMap["foo"] // safe, ok=false if not found

value := myInterface.(string)     // unsafe, panics if not string
value, ok := myInterface.(string) // safe, ok=false if not string
```

Reply:
What would we gain from this? What are the revelant scenarios?

Reply:
To honor defensive programming styles and not only supporting the happy path

Reply:
Fixed in pair-prog session